From fbd7beb99c948c8f5d68df38bc4e266a774d2fa6 Mon Sep 17 00:00:00 2001
From: Marian Buschsieweke <marian.buschsieweke@ovgu.de>
Date: Mon, 10 Oct 2022 14:02:33 +0200
Subject: [PATCH 3/4] components/spi_flash: fix conflicting declarations

`IRAM_ATTR` cannot be used both in the forward declaration and in the
implementation of a function, as this would result in conflicting
attributes due to `__COUNTER__` (used by `IRAM_ATTR`) having different
values in the forward declaration and the implementation.
---
 components/spi_flash/cache_utils.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/components/spi_flash/cache_utils.c b/components/spi_flash/cache_utils.c
index d7d750ab..9ea177e1 100644
--- a/components/spi_flash/cache_utils.c
+++ b/components/spi_flash/cache_utils.c
@@ -62,8 +62,12 @@ static __attribute__((unused)) const char *TAG = "cache";
 #define DPORT_CACHE_GET_VAL(cpuid) (cpuid == 0) ? DPORT_CACHE_VAL(PRO) : DPORT_CACHE_VAL(APP)
 #define DPORT_CACHE_GET_MASK(cpuid) (cpuid == 0) ? DPORT_CACHE_MASK(PRO) : DPORT_CACHE_MASK(APP)
 
-static void IRAM_ATTR spi_flash_disable_cache(uint32_t cpuid, uint32_t *saved_state);
-static void IRAM_ATTR spi_flash_restore_cache(uint32_t cpuid, uint32_t saved_state);
+/* IRAM_ATTR cannot be used in the forward declaration *and* the due to the use
+ * of __COUNTER__ in the IRAM_ATTR definition: This would cause conflicting
+ * section definitions as __COUNTER__ will have a different value in the forward
+ * declaration compared to the implementation. */
+static void /* IRAM_ATTR */ spi_flash_disable_cache(uint32_t cpuid, uint32_t *saved_state);
+static void /* IRAM_ATTR */ spi_flash_restore_cache(uint32_t cpuid, uint32_t saved_state);
 
 static uint32_t s_flash_op_cache_state[2];
 
-- 
2.38.0

